Common mistakes in React (#1550)
enkidevs/curriculum: javascript/react/common-mistakes-in-react/current-state-and-next-state.md

---
author: kapnobatai136
  
aspects:

  - workout

  - deep

type: normal

category: tip

---

# Current state and next state

---
## Content

> In the previous insight we have discussed how the `setState()` update call is asynchronous. For this reason, when you update the state to a value that is based on the old state you might end up with weird values if you do the following:

Reviewer: ```suggestion
In the previous insight we have discussed how the `setState()` update call is asynchronous. For this reason, when you update the state to a value that is based on the old state you might end up with stale values if you do the following:
```

```js
increaseAge = age => {
  this.setState({

>     age: this.state.age + 1

Reviewer: ```suggestion
    // this.state.age might be stale
    age: this.state.age + 1 
```

  });
};
```

> This code relies on the current value of state to update it, but it is not correct. The proper way of updating the state is by using the functional form of `setState`:

Reviewer: ```suggestion
This code relies on the current value of state to update it, which might contain old data. To reliably update the state based on its previous value is to use the functional form of `setState`:
```

```js
increaseAge = () => {
  this.setState(state => ({
    age: state.age + 1
  }));
};
```

> This way, the latest state value is used. When using the functional form of `setState`, you can also pass a second argument - `props` at the time the update is applied. This argument can be used in a manner similar to state.

Reviewer: ```suggestion
This way, the latest state value is **always** provided as an argument to the function you pass into `setState`, and the new state value will become whatever you return from said function. 

Note: the function also receives the latest `props`, as its second argument.
```

If using functional components, the same logic applies to the `useState` hook:

```js
const increaseAge = () => {
  setAge(age => age + 1);
};
```

---
## Practice

> Which function would correctly update the state?

Reviewer: ```suggestion
Which function would **always** update the state to the expected value?
```
